nixos/modules/services/web-apps/keycloak.nix

```
@@ -0,0 +1,441 @@
+            current_owner=$($PSQL -tAc "SELECT pg_catalog.pg_get_userbyid(datdba) FROM pg_catalog.pg_database WHERE datname = 'keycloak'")
+            if [[ "$current_owner" != "${cfg.databaseUsername}" ]]; then
+                $PSQL -tAc 'ALTER DATABASE "keycloak" OWNER TO "${cfg.databaseUsername}"'
+                if [[ -e "${config.services.postgresql.dataDir}/.reassigning_keycloak" ]]; then
+                    echo "Reassigning ownership of database keycloak to user ${cfg.databaseUsername} failed on last boot. Failing..." >&2
+                    exit 1
+                fi
+                touch "${config.services.postgresql.dataDir}/.reassigning_keycloak"
+                $PSQL "keycloak" -tAc "REASSIGN OWNED BY \"$current_owner\" TO \"${cfg.databaseUsername}\""
+                rm "${config.services.postgresql.dataDir}/.reassigning_keycloak"
+            fi
+          '';
+        };
+
+        systemd.services.keycloak = {
+          after = lib.optionals databaseActuallyCreateLocally [
+            "keycloakDatabaseInit.service" "postgresql.service"
+          ];
+          bindsTo = lib.optionals databaseActuallyCreateLocally [
+            "keycloakDatabaseInit.service" "postgresql.service"
+          ];
+          wantedBy = [ "multi-user.target" ];
+          environment = {
+            JBOSS_LOG_DIR = "/var/log/keycloak";
+            JBOSS_BASE_DIR = "/var/lib/keycloak";
+            JBOSS_MODULEPATH = "${cfg.package}/modules";
+          };
+          serviceConfig = {
+            ExecStartPre = systemdPrefixScript "startPre" "+" (''
+              set -eu
+
+              if ! [[ -d /var/lib/keycloak/configuration ]]; then
+                  cp -r ${cfg.package}/standalone/configuration /var/lib/keycloak/
+              fi
+
+              if ! [[ -e ${cfg.databasePasswordFile} ]]; then
```

As above, I would prefer declarative [ConditionPathExists](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Conditions%20and%20Asserts) over imperative `bash` statements.

---

I changed this for the `keycloak` unit, but I'm not sure that it's an improvement - instead of getting an error when the file doesn't exist you now get a silent error that looks like a success.

---

Reverted to checking with `bash`, since it makes runtime errors more explicit.

---

@talyz unfortunate, but that makes sense to me. I'll ping some `systemd` keeners like @andi, @flokli, @arianvp, etc... just to make sure they don't have any feedback I've missed on `ConditionPathExists`.

---

I found a solution - for every `ConditionX` there's also an `AssertX` which fails the unit instead of skipping it. It still doesn't give great error output: it only logs `keycloak.service: Starting requested but asserts failed.`, but if you check the status of the unit you can see the failed condition, so still much better than just skipping it.

---

Why do we need to assert in the first place?  Doesn't keycloak already crash with a happy error message if these files are missing?

---

Well, the initial reason for adding them were to get a better error message than
```
/nix/store/ad05hhpmg2mhn6jqpdgpn1c8xllqdgmk-unit-script-keycloakDatabaseInit-start/bin/keycloakDatabaseInit-start: line 6: /run/keys/db_password: No such file or directory
```
but I suppose that might actually be less cryptic than the assert message.

---

I removed the asserts. If any of the files are configured incorrectly keycloak won't be started and the error message will be pretty clear.